### PR TITLE
Add ability to set event UserIndex and Credential in SDK and add to Linux Lock app.

### PR DESCRIPTION
**Pull request by krypton36**
Info:
SDK does not support the ability for a Lock-App to set the event information for the required data and is hard coded [here](https://github.com/project-chip/connectedhomeip/blob/master/src/app/clusters/door-lock-server/door-lock-server.cpp#L125). This PR updates the SDK to allow setting the UserIndex and Credentials as parameters when setting the lock state as per [spec](https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/app_clusters/DoorLock.adoc#53-lockoperation-event). 

Summary:
- Update SDK to allow setting params
- Update Linux app to send params
- Default parameters set so others may not need to include.
- Validated event sends data.

### CODE REVIEW

**Review by bzbarsky-apple — APPROVED** [review]
Thank you!

File: examples/lock-app/linux/src/LockEndpoint.cpp [hunk 1]
```diff
@@ -457,8 +458,11 @@ bool LockEndpoint::setLockState(DlLockState lockState, const Optional<chip::Byte
         "Lock App: specified PIN code was found in the database, setting door lock state to \"%s\" [endpointId=%d,userIndex=%u]",
         lockStateToString(lockState), mEndpointId, userIndex);
 
-    mLockState = lockState;
-    DoorLockServer::Instance().SetLockState(mEndpointId, mLockState);
+    mLockState                         = lockState;
+    LockOpCredentials userCredential[] = { { CredentialTypeEnum::kPin, uint16_t(credentialIndex) } };
+    auto userCredentials               = MakeNullable<List<const LockOpCredentials>>(userCredential);
+    DoorLockServer::Instance().SetLockState(mEndpointId, mLockState, OperationSourceEnum::kUnspecified,
```
Comment (jrhees-cae) [inline examples/lock-app/linux/src/LockEndpoint.cpp:464]: Calling SetLockState with this signature will have a different effect than the original code. The original code called SetLockState(chip::EndpointId endpointId, DlLockState newLockState) which *only* sets the attribute and doesn't generate any events. The proposed code here will generate a LockOperation event, which will result in the emission of duplicate events when the lock is operated via Matter (not locally). The signature used here will emit an event if the source!=kRemote. So, ideally this code needs to change to specify the source, in which case the userIndex and credential are only used when the source is local.

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 2]
```diff
@@ -107,7 +107,8 @@ bool DoorLockServer::SetLockState(chip::EndpointId endpointId, DlLockState newLo
     return SetAttribute(endpointId, Attributes::LockState::Id, Attributes::LockState::Set, newLockState);
 }
 
-bool DoorLockServer::SetLockState(chip::EndpointId endpointId, DlLockState newLockState, OperationSourceEnum opSource)
+bool DoorLockServer::SetLockState(chip::EndpointId endpointId, DlLockState newLockState, OperationSourceEnum opSource,
+                                  const Nullable<uint16_t> & userIdex, const Nullable<List<const LockOpCredentials>> & credentials)
```
Comment (jrhees-cae) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:111]: nit: change UserIdex to UserIndex

**Review by jrhees-cae — CHANGES_REQUESTED** [review]

File: src/app/clusters/door-lock-server/door-lock-server.cpp (hunk 2 above)
Comment (krypton36) [reply src/app/clusters/door-lock-server/door-lock-server.cpp:111]: Fixed. Thank you.

File: examples/lock-app/linux/src/LockEndpoint.cpp (hunk 1 above)
Comment (krypton36) [reply examples/lock-app/linux/src/LockEndpoint.cpp]: Good Catch! Validated that there was two events and now only a single one. Updated, can you please review again?

File: examples/lock-app/linux/src/LockEndpoint.cpp [hunk 3]
```diff
@@ -457,8 +459,10 @@ bool LockEndpoint::setLockState(DlLockState lockState, const Optional<chip::Byte
         "Lock App: specified PIN code was found in the database, setting door lock state to \"%s\" [endpointId=%d,userIndex=%u]",
         lockStateToString(lockState), mEndpointId, userIndex);
 
-    mLockState = lockState;
-    DoorLockServer::Instance().SetLockState(mEndpointId, mLockState);
+    mLockState                         = lockState;
+    LockOpCredentials userCredential[] = { { CredentialTypeEnum::kPin, uint16_t(credentialIndex) } };
+    auto userCredentials               = MakeNullable<List<const LockOpCredentials>>(userCredential);
+    DoorLockServer::Instance().SetLockState(mEndpointId, mLockState, opSource, Nullable<uint16_t>(userIndex + 1), userCredentials);
```
Comment (bzbarsky-apple) [inline examples/lock-app/linux/src/LockEndpoint.cpp:465]:
```suggestion
    DoorLockServer::Instance().SetLockState(mEndpointId, mLockState, opSource, MakeNullable(static_cast<uint16_t>(userIndex + 1)), userCredentials);
```

**Review by jrhees-cae — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]

**Pull request merged**